### DISCUSSION TITLE
Questions about the MCA

### DISCUSSION THREAD

**Discussion by Kim1230** [verificarlo/verificarlo#328] [Q&A] [answered]
Hello, I read about the paper about the verificarlo tutorial and a full book High Performance Computing code optimizations: Tuning performance and accuracy. And  I have a question and hope your answer. 
![image](https://user-images.githubusercontent.com/107012598/204173355-47bc2509-2cda-42dd-92b5-6a9eb86e10d3.png)
Why it has to divide x to y and z？ 
![image](https://user-images.githubusercontent.com/107012598/204173456-4f2fed8e-ebf8-42cf-9420-cc7ecc48d0d7.png)
what is the meaning of the (ex-ez)？ and if the extended precision is 2p in verificarlo. I read the code of the mca, it shows that 
![image](https://user-images.githubusercontent.com/107012598/204173543-1836cbb8-888e-4c39-9ecf-e36cc17c8e0e.png). It doesn't equal to the 2p? 

Hope for your answer. Thanks!

**Comment by pablooliveira** [comment, marked as answer]
Hi @Kim1230.

Regarding the first question " Why it has to divide x yo y and z ? ", I'm not sure exactly what you mean. But in the example you quote, $x = y + z$ and we are operating in RR mode; so $inexact(x)$ is the summation of $y$ and $z$ and of the stochastic noise.
The stochastic noise is scaled by a factor $2^{e_x-t}$, so it simulates a virtual precision $t$ in the mantissa of the result $x$.
The problem here, is that the smaller number $z$ is then shifted to the right by $e_z-e_x$ bits.

Regarding the second questions on the extended precision. The precision of a float is 24 bits and a double 53 bits. In verificarlo we need indeed at least $2p$ for the extended computations with the noise.  For performance reasons, we use existing extended types such as double (to extend float) and quad (to extend double). Thus  we have $24\times 2 < 53$ and $53\times 2 < 112$. We indeed have at least $2p$ bits for the extended computations.

**Reply by Kim1230** [reply]
Hello@pablooliveira, Thanks for your response
The first question I mentioned is because I see the inexact(x) has 2 representations。 First, I saw it will be :
![image](https://user-images.githubusercontent.com/107012598/204234547-f9526e00-f326-45de-8f73-ae3dd125f59c.png)
And I thought MCA is considering the stochastic noise to the x. But in verificarlo, furthermore, it changes x to y plus z. I don't know why we don't directly simulates a virtual precision t in the mantissa of the result x, but to have z and e_z-e_x bits here? To summrize, I don't the meaning of z and why it is used?
Hope for your answer

**Reply by pablooliveira** [reply]
Yes $inexact(x) = x + 2^{e_x-t}\xi$ . But in the example you are quoting, we consider $x = y + z$ and $e_x = e_y$ ; so we want to compute the MCA result of a summation between $y$ and $z$, where the result $x$ has the same exponent than one of the summands, here $y$.

By replacing $x$ by $y+z$ and factoring the expression with $2^{e_x}$, the exponent of the result, we get the right-hand side.

**Reply by Kim1230** [reply]
So, it is two situation? One is considering  MCA result of the solo variable x and the other is considering the summation equation y+z?

**Reply by pablooliveira** [reply]
Yes. But both equations are coherent. We just developed the equation by injecting y+z to see the effect on each summand's mantissa.
